src/core/fs/romfs.cpp

```diff
@@ -0,0 +1,136 @@
+        path += std::string(node.name.begin(), node.name.end()) + "/";
+        indentation++;
+        for (auto& directory : node.directories) {
+            printNode(*directory, indentation, path);
+        }
+        indentation--;
+    }
+
+    std::vector<std::unique_ptr<RomFSNode>> parseDirectory(const uintptr_t metadataBase, const uintptr_t metadataOffset) {
+        std::vector<std::unique_ptr<RomFSNode>> directories {};
+
+        // Get offset of first child directory
+        u32* metadataPtr = (u32*)(metadataBase + metadataOffset);
+        metadataPtr += 2;
+        u32 currentDirectoryOffset = *metadataPtr;
+
+        // Loop over all the sibling directories of the first child to get all the children directories
+        while (currentDirectoryOffset != metadataInvalidEntry) {
+            metadataPtr = (u32*)(metadataBase + currentDirectoryOffset);
+            metadataPtr++; // Skip the parent offset
+            u32 siblingDirectoryOffset = *metadataPtr++;
+            metadataPtr++; // Skip the child offset
+            metadataPtr++; // Skip the first file offset
+            metadataPtr++; // Skip the next directory in hash table offset
+            u32 nameLength = (*metadataPtr++) / 2;
+
+            // Arbitrary limit
+            if (nameLength > 128) {
+                printf("Invalid directory name length: %08X\n", nameLength);
+                return {};
+            }
+
+            char16_t* namePtr = (char16_t*)metadataPtr;
+            std::u16string name(namePtr, nameLength);
+
+            std::unique_ptr directory = std::make_unique<RomFSNode>();
+            directory->isDirectory = true;
+            directory->name = name;
+            directory->offset = currentDirectoryOffset;
+            directories.push_back(std::move(directory));
+
+            currentDirectoryOffset = siblingDirectoryOffset;
+        }
+
+        // Loop over all the children directories to get their children
+        for (auto& directory : directories) {
+            directory->directories = parseDirectory(metadataBase, directory->offset);
+        }
+
+        return directories;
+    }
+
+    std::unique_ptr<RomFSNode> parseRomFSTree(uintptr_t romFS, u64 romFSSize) {
+        IVFC::IVFC ivfc;
+        size_t ivfcSize = IVFC::parseIVFC((uintptr_t)romFS, ivfc);
+
+        if (ivfcSize == 0) {
+            printf("Failed to parse IVFC\n");
+            return {};
+        }
+
+        uintptr_t masterHashOffset = RomFS::align(ivfcSize, 0x10);
+        // For a weird reason, the level 3 offset is not the one in the IVFC, instead it's
+        // the first block after the master hash
+        // TODO: Find out why and explain in the comment
+        uintptr_t level3Offset = RomFS::align(masterHashOffset + ivfc.masterHashSize, ivfc.levels[2].blockSize);
+        uintptr_t const level3Base = (uintptr_t)romFS + level3Offset;
+        u32* level3Ptr = (u32*)level3Base;
```

Comment:
I didn't even know that placement for const is legal
![AW4183505_07](https://github.com/wheremyfoodat/Panda3DS/assets/44909372/5c12206d-0468-4af8-b2f9-1a0d8d4516dd)